cpp/ql/src/Security/CWE/CWE-843/TypeConfusion.ql

```diff
@@ -0,0 +1,228 @@
+      this instanceof StaticCast
+      or
+      this instanceof ReinterpretCast
+    ) and
+    toType = this.getExplicitlyConverted().getUnspecifiedType().stripType() and
+    not this.isImplicit()
+  }
+
+  Class getConvertedType() { result = toType }
+
+  bindingset[this, t]
+  pragma[inline_late]
+  predicate compatibleWith(Type t) {
+    t.stripType() = this.getConvertedType()
+    or
+    prefix(this.getConvertedType(), t.stripType())
+    or
+    prefix(t.stripType(), this.getConvertedType())
+    or
+    t.stripType().(Class).getABaseClass+() = this.getConvertedType()
+    or
+    t.stripType() = this.getConvertedType().getABaseClass+()
+  }
+}
+
+/**
+ * Holds if `source` is an allocation that allocates a value of type `state`.
+ */
+predicate isSourceImpl(DataFlow::Node source, Class state) {
+  state = source.asExpr().(AllocationExpr).getAllocatedElementType().stripType()
+}
+
+module RelevantStateConfig implements DataFlow::ConfigSig {
+  predicate isSource(DataFlow::Node source) { isSourceImpl(source, _) }
+
+  predicate isBarrier(DataFlow::Node node) {
+    // We disable flow through global variables to reduce FPs from infeasible paths
+    node instanceof DataFlow::VariableNode
+  }
+
+  predicate isSink(DataFlow::Node sink) {
+    exists(UnsafeCast cast | sink.asExpr() = cast.getUnconverted())
+  }
+}
+
+module RelevantStateFlow = DataFlow::Global<RelevantStateConfig>;
+
+predicate relevantState(DataFlow::Node sink, Class state) {
+  exists(DataFlow::Node source |
+    RelevantStateFlow::flow(source, sink) and
+    isSourceImpl(source, state)
+  )
+}
+
+predicate isSinkImpl(DataFlow::Node sink, Class state, Type convertedType, boolean compatible) {
+  exists(UnsafeCast cast |
+    relevantState(sink, state) and
+    sink.asExpr() = cast.getUnconverted() and
+    convertedType = cast.getConvertedType()
+  |
+    if cast.compatibleWith(state) then compatible = true else compatible = false
+  )
+}
+
+module BadConfig implements DataFlow::StateConfigSig {
+  class FlowState extends Class {
+    FlowState() { isSourceImpl(_, this) }
+  }
+
+  predicate isSource(DataFlow::Node source, FlowState state) { isSourceImpl(source, state) }
+
+  predicate isBarrier(DataFlow::Node node) { RelevantStateConfig::isBarrier(node) }
+
+  predicate isSink(DataFlow::Node sink, FlowState state) { isSinkImpl(sink, state, _, false) }
+
+  predicate isBarrierOut(DataFlow::Node sink, FlowState state) { isSink(sink, state) }
+}
+
+module BadFlow = DataFlow::GlobalWithState<BadConfig>;
+
+module GoodConfig implements DataFlow::StateConfigSig {
+  class FlowState = BadConfig::FlowState;
+
+  predicate isSource(DataFlow::Node source, FlowState state) { BadConfig::isSource(source, state) }
+
+  predicate isBarrier(DataFlow::Node node) { BadConfig::isBarrier(node) }
+
+  predicate isSink(DataFlow::Node sink, FlowState state) {
+    isSinkImpl(sink, state, _, true) and
+    BadFlow::flowTo(sink)
+  }
+}
+
+module GoodFlow = DataFlow::GlobalWithState<GoodConfig>;
```

Comment:
I've added some QLDoc in 035c405c0b1cc04bce84eff6d958d9a919d3ff24 to explain this. The TLDR is that `BadFlow` (and `GoodFlow`) needs to compare the type of the allocation (i.e., the one in `isSource`) to the type at the cast site (i.e., the one in `isSink`). And in order to know the source type in the implementation of `isSink` we need to use a flow state. However, if you have no way to relate the sink type to the flow state then you get a cartesian product in `isSinkImpl`:
```ql
predicate isSinkImpl(DataFlow::Node sink, Class state, Type convertedType, boolean compatible) {
  exists(UnsafeCast cast |
    // imagine this line isn't here
    // relevantState(sink, state) and
    sink.asExpr() = cast.getUnconverted() and
    convertedType = cast.getConvertedType()
  |
    // this is now a cartesian product then
    if cast.compatibleWith(state) then compatible = true else compatible = false
  )
```

Reply:
I can see that `RelevantStateFlow` is an optimization for `GoodFlow` and `BadFlow`, what I'm not sure of is why / whether we still need to keep `GoodFlow` and `BadFlow` once we have it.  Once we've established flow from a source to a sink with `RelevantStateFlow`, what details do we need to check with a second pass that aren't available from just the source + sink anyway?  I can, for example, replace `BadFlow` in the query with `RelevantStateFlow` and it does not seem to affect results (on tests, anyway).

Reply:
You can distinguish flow from `new Cat` vs flow from `new Dog` by looking back at the source after the flow is complete - you don't _need_ a flow state unless there's something interesting happening mid-flow.  You do reference `state` in `isBarrierOut`, so it could be something subtle is happening there???

Reply:
> You can distinguish flow from `new Cat` vs flow from `new Dog` by looking back at the source after the flow is complete

That will generate a lot of paths, though. With the current approach, we're only generating paths from an allocation to an invalid cast. So I'm more less concerned about the performance of the query as-is.

> You do reference state in `isBarrierOut`, so it could be something subtle is happening there???

Yeah, that's a good point: we don't want multiple alerts as soon as there's a single invalid cast along the path. But we can't just block flow through _any_ cast (since not all casts are bad). So we need to know the type of the allocation to deduce whether the cast is bad. And for that we need the flow state.

Reply:
I ended up removing both `GoodFlow` and `BadFlow` in 7b0df57d7a485d71cdb5f060dccc2388590674bf to fix the performance problem on ChakraCore. So now we're doing what you suggested above in:

> You can distinguish flow from `new Cat` vs flow from `new Dog` by looking back at the source after the flow is complete [...]

And as expected, we now see more paths being generated that doesn't actually contribute to any alert. We also cannot block flow out of invalid casts now which is a bit unfortunate. But that seems to be what's necessary to fix performance since that now looks much better 🎉